src/lir/converter/mod.rs

```diff
@@ -0,0 +1,240 @@
+// SPDX-License-Identifier: Apache-2.0
+use crate::codegen::cfg::BasicBlock;
+use crate::lir::{Block, LIR};
+use crate::{
+    codegen::{
+        self,
+        cfg::{self, ControlFlowGraph},
+    },
+    sema::ast::{self, Namespace, Parameter, RetrieveType},
+};
+
+use super::{
+    expressions::Operand,
+    instructions::Instruction,
+    lir_type::{InternalCallTy, Type},
+    vartable::Vartable,
+};
+
+mod expression;
+mod instruction;
+mod lir_type;
+mod vartable;
+
+/// A Converter converts a ControlFlowGraph into a Lower Intermediate Representation.
+pub struct Converter<'a> {
+    /// a reference to the Namespace is used to retrieve useful information like enum types, address length, etc.
+    ns: &'a Namespace,
+    /// a reference to the ControlFlowGraph is used to retrieve the instructions.
+    cfg: &'a ControlFlowGraph,
+}
+
+impl<'input> Converter<'input> {
+    pub fn new(ns: &'input Namespace, cfg: &'input ControlFlowGraph) -> Self {
+        Self { ns, cfg }
+    }
+
+    pub fn fn_selector_length(&self) -> u8 {
+        self.ns.target.selector_length()
+    }
+
+    pub fn address_length(&self) -> usize {
+        self.ns.address_length
+    }
+
+    pub fn convert_user_type(&self, user_ty: &ast::Type) -> Type {
+        // clone happens here because function unwrap_user_type takes ownership
+        let real_ty = user_ty.clone().unwrap_user_type(self.ns);
+        self.lowering_ast_type(&real_ty)
+    }
+
+    pub fn convert_enum_type(&self, enum_no: usize) -> Type {
+        let ty = &self.ns.enums[enum_no].ty;
+        self.lowering_ast_type(ty)
+    }
+
+    pub fn value_length(&self) -> usize {
+        self.ns.value_length
+    }
+
+    pub fn get_ast_type_by_id(&self, id: &usize) -> ast::Type {
+        match self.cfg.vars.get(id) {
+            Some(var) => var.ty.clone(),
+            None => panic!("Cannot find type for id {}", id),
+        }
+    }
+
+    pub fn to_operand(
+        &self,
+        expr: &codegen::Expression,
+        vartable: &mut Vartable,
+    ) -> Option<Operand> {
+        match expr {
+            codegen::Expression::NumberLiteral { ty, value, loc, .. } => {
+                let ssa_ty = self.lowering_ast_type(ty);
+                Some(Operand::new_number_literal(value, ssa_ty, *loc))
+            }
+            codegen::Expression::BoolLiteral { value, loc, .. } => {
+                Some(Operand::new_bool_literal(*value, *loc))
+            }
+            codegen::Expression::Variable { loc, var_no, .. } => {
+                Some(Operand::new_id(*var_no, *loc))
+            }
+            codegen::Expression::FunctionArg { loc, arg_no, .. } => {
+                vartable.get_function_arg(*arg_no, *loc)
+            }
+            _ => None,
+        }
+    }
+
+    pub fn to_operand_and_insns(
```

Comment:
Usually, `pub` function should come with a doc comment. Here it's not too big of an issue because the name and argument name and types make it somewhat obvious. But still, there is `fn to_operand`, `fn to_operand_and_insns`, `fn to_operand_option_and_insns` etc.

Some doc comments explaining the higher level workings and your thoughts here would be appreciated.